design/rfc-01-fwu-suit/rfc-01.rst

```diff
@@ -0,0 +1,693 @@
+   *  It requires the Update service to be re-entrant, when transferring the requested payload to the service.
+   *  On deployments where the Update service is isolated from the Update client, the callback mechanism will only operate in the client, and still require a state-based approach within the service.
+
+   Therefore, a callback-based API would not be usable in many systems. In order to define a single API for this feature that can be used in all types of system execution environment, the chosen approach for fetching payloads is to use a return value from manifest processing that signals the need for the application to fetch and transfer a payload before resuming the SUIT manifest processing.
+
+Complex installation sequences
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Where a SUIT manifest includes additional payload fetching as part of the suit-install command sequence, the manifest processing and payload fetching cycle would be repeated after the installation commences. This might occur immediately if no reboot is required, or as part of a dedicated Bootloader/Installer state of the system if this requires a system restart.
+
+`Figure 2 <fig-install-flow_>`_ shows an example of the additional flow when fetching during installation after a reboot.
+
+.. _fig-install-flow:
+
+.. figure:: suit-install.svg
+
+   **Figure 2** *The additional flow in a complex SUIT installation*
+
+
+Draft API design
+----------------
+
+Envelope and payload transfer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`Figure 3 <fig-fetch_>`_ shows the detailed call sequence for the Update client when initially processing a SUIT envelope
+
+.. _fig-fetch:
+
+.. figure:: fetch-sequence.svg
+
+   **Figure 3** *The normal call sequence for initial SUIT update*
+
+   This shows the transfer of the envelope, and the processing of the suit-payload-fetch command sequence.
+
+The detailed steps of the flow are as follows (the numbers refer to the corresponding number in Figure 3):
+
+*  1: The process is started by transferring the SUIT envelope as a firmware image using a component identifier allocated to the SUIT envelope component.
+
+*  6: The call to ``psa_fwu_finish()`` behaves differently when processing a SUIT envelope. On a successful transfer, the call will return a new response code, ``PSA_FWU_PROCESSING_REQUIRED``, to indicate that the component requires processing. At this point the envelope component will be in a new ``PSA_FWU_FETCHING`` state, instead of the typical ``PSA_FWU_CANDIDATE`` state.
+
+*  7: If processing is required, the Update client then calls ``psa_fwu_process()`` to begin manifest processing. At this point the Update service will do the following:
+
+   * Verify and authenticate the manifest.
+   * Process the Update command sequences: system validation, dependency resolution, payload fetch, payload verification.
```

Comment:
There is no dedicated ‘system validation’ sequence defined in SUIT specification. Maybe it is worth to remove it?

Reply:
I think when I first wrote this I was thinking about checking that the update matches the system (device type, etc.) - but that would be integrated into one of the other sequences.

Fixed.